### PR TITLE
Scatter Simulation in SIRF

### PR DESCRIPTION
**Pull request by NikEfth**
Work in progress. 
In order to be able to run this branch you have to switch to the Scatter_simulation branch in STIR (recompile and install). 
You can get this branch from [here](https://github.com/NikEfth/STIR/tree/Scatter_simulation).

Progress: 
- [x] We can create a Single Scatter simulation object from SIRF
- [x] Set the template projdata info in SSS and check
- [ ] Run ```SingleScatterSimulator::set_up()```
- [ ] Run ```SingleScatterSimulator::simulate()```

### CODE REVIEW

**Comment by KrisThielemans** [comment]
thanks. Ideally, you modify the travis.yml to use the SuperBuild with appropriate `STIR_TAG` and `STIR_URL`. otherwise, please add `[ci skip]` to commit messages, as we'd get a lot of failures...

File: examples/Python/PET/scatter_simulation.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+    #
+    # # z-pixel coordinate of the xy-crossection to show
+    # z = int(image_size[0]/2)
+    #
+    # # show the phantom image
+    # image_array = image.as_array()
+    # show_2D_array('Phantom image', image_array[z,:,:])
+
+    # Create a template Acquisition Model
+    tmpl_acq_data = pystir.cSTIR_acquisitionsDataFromScannerInfo('ECAT 931', 1, 0, 1)
```
Comment (rijobro) [inline examples/Python/PET/scatter_simulation.py]: This line will obviously have to be replaced before the PR gets merged.

Comment (NikEfth) [reply examples/Python/PET/scatter_simulation.py]: I see that in other examples you read the ProjDataInfo from a file. Is there another way? 
How would you change it?

Comment (rijobro) [reply examples/Python/PET/scatter_simulation.py]: Which other examples does it exist in? Our python wrapping code should be the only code calling anything with `pystir`. 

https://github.com/CCPPETMR/SIRF/blob/ea4a56f9f07ae5a17d74887fbca002dc69313b5c/src/xSTIR/pSTIR/STIR.py#L551-L583

The attached code snipped shows that the constructor of `AcquisitionData` accepts a scanner name as input and calls `cSTIR_acquisitionsDataFromScannerInfo` itself. So I think you should simply be able to use: 

```
tmpl_acq_data = AcquisitionData('ECAT 931', 1, 0, 1)
```

File: src/xSTIR/pSTIR/STIR.py [hunk 2]
```diff
@@ -237,6 +243,76 @@ def get_radii(self):
...
+        if not self.filename:
+            self.handle = pystir.cSTIR_newObject(self.name)
+        else:
+            self.handle = pystir.cSTIR_objectFromFile
+
+        check_status(self.handle)
+
+    def __del__(self):
+        if self.handle is not None:
+            pyiutild.deleteDataHandle(self.handle)
```
Comment (rijobro) [inline src/xSTIR/pSTIR/STIR.py]: This is the problematic line. Check spelling.

File: src/xSTIR/pSTIR/STIR.py [hunk 3]
```diff
@@ -237,6 +243,76 @@ def get_radii(self):
...
+
+    def set_up(self):
+        print('SingleScatterSimulator:: I am in set up')
+
+    def set_Acquisition_template(self, arg):
+        ''' Set the template of the acquisition data.
+        '''
+        par = pystir.cSTIR_getAcquisitionsData(arg.handle, arg.name);
+        _setParameter(self.handle, self.name, 'template_proj_data_info', par)
```
Comment (rijobro) [inline src/xSTIR/pSTIR/STIR.py]: Delete first line. For second line, replace `par` with `arg.handle`.

File: src/xSTIR/pSTIR/STIR.py [hunk 4]
```diff
@@ -237,6 +243,76 @@ def get_radii(self):
...
+
+    def set_low_energy_threshold(self, threshold):
```
Comment (KrisThielemans) [inline src/xSTIR/pSTIR/STIR.py]: should be in `STIRAcquisitionData` (if we really need them)

**Comment by NikEfth** [comment]
Should I expose the downsampling functions?

**Comment by KrisThielemans** [comment]
> Should I expose the downsampling functions?

I don't think so. I think the SIRF user wants to be able to get scatter estimate out in the same size as measured data. That will do all the down/up sampling in one go. Potentially they want to play with the parameters, but I would do that initially by telling them to use a STIR parameter file (which is what we do with the recon as well). Otherwise we get into a load of code duplication between SIRF and STIR.

**Comment by NikEfth** [comment]
> That will do all the down/up sampling in one go
So should I downsample something like 50% by default?

**Comment by KrisThielemans** [comment]
default down-sampling is a STIR issue, not a SIRF one :-) see https://github.com/UCL/STIR/pull/44#issuecomment-497723931

File: src/xSTIR/pSTIR/STIR.py (hunk 2 above)
Comment (NikEfth) [reply src/xSTIR/pSTIR/STIR.py]: Corrected. thank you

File: examples/Python/PET/scatter_simulation.py (hunk 1 above)
Comment (NikEfth) [reply examples/Python/PET/scatter_simulation.py]: :ok:

File: src/xSTIR/pSTIR/STIR.py (hunk 3 above)
Comment (NikEfth) [reply src/xSTIR/pSTIR/STIR.py]: :ok:

File: src/xSTIR/pSTIR/STIR.py [hunk 5]
```diff
@@ -237,6 +243,76 @@ def get_radii(self):
...
+        check_status(self.handle)
+
+    def __del__(self):
+        if self.handle is not None:
+            pyiutild.deleteDataHandle(self.handle)
+
+    def set_up(self):
+        print('SingleScatterSimulator:: I am in set up')
+
+    def set_Acquisition_template(self, arg):
```
[the remark this answers is not in the document]
Comment (NikEfth) [reply src/xSTIR/pSTIR/STIR.py]: :ok:

File: src/xSTIR/pSTIR/STIR.py (hunk 4 above)
Comment (NikEfth) [reply src/xSTIR/pSTIR/STIR.py]: For the time I moved them there

File: src/xSTIR/pSTIR/STIR.py [hunk 6]
```diff
@@ -237,6 +243,76 @@ def get_radii(self):
...
+
+    def set_high_energy_threshold(self, threshold):
+        _set_float_par(self.handle, self.name, 'high_energy_threshold', threshold)
+
+    def set_energy_resolution(self, energy_resolution):
+        ''' Set the energy properties of the Scanner in keV.
+        '''
+        _set_float_par(self.handle, self.name, 'set_Scanner_energy_information', energy_resolution)
+
+    def has_Acquisition_template(self):
```
[the remark this answers is not in the document]
Comment (NikEfth) [reply src/xSTIR/pSTIR/STIR.py]: :ok:

File: examples/Python/PET/scatter_simulation.py (hunk 1 above)
Comment (evgueni-ovtchinnikov) [reply examples/Python/PET/scatter_simulation.py]: You absolutely must have it the way Richard suggested - `pystir` objects are not to be used directly.

File: src/xSTIR/pSTIR/STIR.py [hunk 7]
```diff
@@ -237,6 +243,38 @@ def get_radii(self):
...
+    def __init__(self, filename = ''):
+        self.handle = None
+        self.image = None
+        self.name = 'SingleScatterSimulation'
+        self.filename = filename
+
+        if not self.filename:
+            self.handle = pystir.cSTIR_newObject(self.name)
+        else:
+            self.handle = pystir.cSTIR_objectFromFile
```
Comment (evgueni-ovtchinnikov) [inline src/xSTIR/pSTIR/STIR.py]: missing arguments `(self.name, self.filename)`

**Comment by KrisThielemans** [comment]
does this need to be abandoned over #404?

**Comment by KrisThielemans** [comment]
@NikEfth is this superseded by #404 ? Please close if so. If not, can you clarify?

**Comment by KrisThielemans** [comment]
closing this as we believe superseded by #404

**Pull request closed without merging**